server/src/test/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingServiceTests.java

```diff
@@ -0,0 +1,693 @@
+            .nodeFeatures(
+                Map.of(
+                    "n1",
+                    Set.of(DataStreamAutoShardingService.DATA_STREAM_AUTO_SHARDING_FEATURE.id()),
+                    "n2",
+                    Set.of(DataStreamAutoShardingService.DATA_STREAM_AUTO_SHARDING_FEATURE.id())
+                )
+            )
+            .metadata(builder)
+            .build();
+
+        {
+            // autosharding disabled
+            DataStreamAutoShardingService disabledAutoshardingService = new DataStreamAutoShardingService(
+                Settings.EMPTY,
+                clusterService,
+                new FeatureService(List.of(new FeatureSpecification() {
+                    @Override
+                    public Set<NodeFeature> getFeatures() {
+                        return Set.of(DataStreamAutoShardingService.DATA_STREAM_AUTO_SHARDING_FEATURE);
+                    }
+                })),
+                System::currentTimeMillis
+            );
+
+            AutoShardingResult autoShardingResult = disabledAutoshardingService.calculate(state, dataStream, 2.0);
+            assertThat(autoShardingResult.type(), is(NOT_APPLICABLE));
+            assertThat(autoShardingResult.coolDownRemaining(), is(TimeValue.MAX_VALUE));
+            assertThat(autoShardingResult.targetNumberOfShards(), is(0));
+        }
+
+        {
+            // cluster doesn't have feature
+            ClusterState stateNoFeature = ClusterState.builder(ClusterName.DEFAULT).metadata(Metadata.builder()).build();
+
+            DataStreamAutoShardingService noFeatureService = new DataStreamAutoShardingService(
+                Settings.builder()
+                    .put(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_ENABLED, true)
+                    .putList(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey(), List.of())
+                    .build(),
+                clusterService,
+                new FeatureService(List.of()),
+                () -> now
+            );
+
+            AutoShardingResult autoShardingResult = noFeatureService.calculate(stateNoFeature, dataStream, 2.0);
+            assertThat(autoShardingResult.type(), is(NOT_APPLICABLE));
+            assertThat(autoShardingResult.coolDownRemaining(), is(TimeValue.MAX_VALUE));
+            assertThat(autoShardingResult.targetNumberOfShards(), is(0));
+        }
+
+        {
+            // patterns are configured to exclude the current data stream
+            DataStreamAutoShardingService noFeatureService = new DataStreamAutoShardingService(
+                Settings.builder()
+                    .put(DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_ENABLED, true)
+                    .putList(
+                        DataStreamAutoShardingService.DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey(),
+                        List.of(dataStreamName + "*")
```

Comment:
Let's make it not the first data stream in the list (just to test things):
```suggestion
                        List.of("foo", dataStreamName + "*")
```